File: strings/ransom_note.rb
# @param {String} ransom_note
# @param {String} magazine
# @return {Boolean}
def can_construct(ransom_note, magazine)
    magazine_chars = Hash.new(0)
    magazine.each_char { |c| magazine_chars[c] += 1 }
    
    ransom_note.each_char do |c|
        return false if magazine_chars[c] == 0
        magazine_chars[c] -= 1
    end
    
    true
end

File: strings/ransom_note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # Create character frequency dictionary
        magazine_chars = {}
        for c in magazine:
            magazine_chars[c] = magazine_chars.get(c, 0) + 1
            
        # Check if we have enough characters
        for c in ransomNote:
            if c not in magazine_chars or magazine_chars[c] == 0:
                return False
            magazine_chars[c] -= 1
            
        return True